### PR TITLE
Support for HPE PDUs

### PR DESCRIPTION
**Pull request by mbuchmann-hpe**
### Summary and Scope

Adds support for HPE PDUs

DOES THIS CHANGE INVOLVE ANY SCHEME CHANGES?  N

REMINDER: HAVE YOU INCREMENTED VERSION NUMBERS? 1.30.0

REMINDER 2: HAVE YOU UPDATED THE COPYRIGHT PER HPE GUIDELINES: (C) Copyright 2014-2021 Hewlett Packard Enterprise Development LP    ? Y

### Issues and Related PRs

* CASMHMS-5238

### Testing

Tested on:
* Tested on Loki with HPE PDU
* Tested on non HPE PDU and appears to continue to work

### Risks and Mitigations

Moderate risk - changes should only affect PDUs

### CODE REVIEW

File: cmd/capmcd/bmcapi.go [hunk 1]
```diff
@@ -315,13 +316,33 @@ func (d *CapmcD) doBmcPowerCall(call bmcCall) bmcPowerRc {
...
-		body = fmt.Sprintf(`{"PowerState": "%s"}`, resetType)
+		var HPEPDU = true
+		if strings.Contains(ni.BmcFQDN, "rts") {
+			HPEPDU = false
+		}
+		if HPEPDU {
+			outletNum := strings.Split(ni.Hostname, "v")
+			if len(outletNum) < 2 {
+				log.Printf("Could not get outlet number")
+				outletNum = append(outletNum, "")
```
Comment (mjendrysik-hpe) [inline cmd/capmcd/bmcapi.go:335]: Not sure what is going on here, why append an empty zero length string twice?

**Review by nieuwsma — CHANGES_REQUESTED** [review]
Temporarily BLOCKED by CASM-2670

File: cmd/capmcd/bmcapi.go [hunk 2]
```diff
@@ -315,13 +316,33 @@ func (d *CapmcD) doBmcPowerCall(call bmcCall) bmcPowerRc {
...
+				outletNum = append(outletNum, "")
+			} else {
+				log.Printf("Outlet number: %s", outletNum)
```
Comment (mjendrysik-hpe) [inline cmd/capmcd/bmcapi.go:338]: I would drop this log.Printf, there is no context.

File: cmd/capmcd/bmcapi.go [hunk 3]
```diff
@@ -315,13 +316,33 @@ func (d *CapmcD) doBmcPowerCall(call bmcCall) bmcPowerRc {
...
 		fallthrough
 	case "CabinetPDUPowerConnector":
-		body = fmt.Sprintf(`{"PowerState": "%s"}`, resetType)
+		var HPEPDU = true
+		if strings.Contains(ni.BmcFQDN, "rts") {
+			HPEPDU = false
+		}
+		if HPEPDU {
+			outletNum := strings.Split(ni.Hostname, "v")
+			if len(outletNum) < 2 {
```
Comment (mjendrysik-hpe) [inline cmd/capmcd/bmcapi.go:333]: If we don't get an outletNum, will the resulting call to the PDU succeed? Should this be an error instead?

File: cmd/capmcd/bmcapi.go [hunk 4]
```diff
@@ -344,11 +365,35 @@ func (d *CapmcD) doBmcPowerCall(call bmcCall) bmcPowerRc {
...
+		req.SetBasicAuth(ni.BmcUser, ni.BmcPass)
+		req.Header.Set("Accept", "*/*")
+		req.Header.Set("Content-Type", "application/json")
+		// execute the request
+		rfClientLock.RLock()
+		rsp, err := d.rfClient.Do(req)
+		rfClientLock.RUnlock()
+		if err != nil {
+			log.Printf("POST %s\n Body           --> %s\n %s Network Error: %s",
+				sessionAuthPath, sessionAuthBody, ni.BmcType, err)
```
Comment (mjendrysik-hpe) [inline cmd/capmcd/bmcapi.go:381]: I don't believe the sessionAuthBody has been sanitized at this point, logging it will expose the credentials.

**Review by mjendrysik-hpe — CHANGES_REQUESTED** [review]

File: cmd/capmcd/bmcapi.go (hunk 2 above)
Comment (mjendrysik-hpe) [reply cmd/capmcd/bmcapi.go:338]: Also, later on the body is logged which will include the outlet number.

File: kubernetes/cray-hms-capmc/Chart.yaml [hunk 5]
```diff
@@ -3,4 +3,4 @@ apiVersion: v1
 description: "Kubernetes resources for cray-hms-capmc"
 name: "cray-hms-capmc"
 home: "HMS/hms-capmc"
-version: 1.29.0
+version: 1.30.0
```
Comment (mjendrysik-hpe) [inline kubernetes/cray-hms-capmc/Chart.yaml:6]: I would just drop this change from the mod since it is going away anyway.

**Review by mjendrysik-hpe — CHANGES_REQUESTED** [review]

File: cmd/capmcd/bmcapi.go (hunk 1 above)
Comment (mbuchmann-hpe) [reply cmd/capmcd/bmcapi.go]: Just returns now if outlet unavailable

File: cmd/capmcd/bmcapi.go (hunk 2 above)
Comment (mbuchmann-hpe) [reply cmd/capmcd/bmcapi.go]: Updated

File: cmd/capmcd/bmcapi.go (hunk 4 above)
Comment (mbuchmann-hpe) [reply cmd/capmcd/bmcapi.go]: Removed body

File: kubernetes/cray-hms-capmc/Chart.yaml (hunk 5 above)
Comment (mbuchmann-hpe) [reply kubernetes/cray-hms-capmc/Chart.yaml]: Removed file

File: cmd/capmcd/bmcapi.go (hunk 3 above)
Comment (mbuchmann-hpe) [reply cmd/capmcd/bmcapi.go:333]: Just returns now if outlet unavailable

**Review by mjendrysik-hpe — APPROVED** [review]

**Review by nieuwsma — APPROVED** [review]

**Pull request merged**